### PR TITLE
Remove some repetition from Greens

### PR DESCRIPTION
**Pull request by geograham**
## Linked Issues

Papercut #205 - three suggestions are made in this issue. The first addresses repetition and is covered by this pull request. The second is now a separate issue (#3273). The third was discussed by Bluemira developers but does not need to be implemented.  

## Description

Calculations that are repeated in multiple locations are now functions.

## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/magnetostatics/greens.py [hunk 1]
```diff
@@ -288,12 +278,44 @@ def greens_dpsi_dz(
...
-    i2 = ellipe_nb(k2) / (a**3 * (1 - k2))
-    return MU_0_2PI * ((z - zc) * (i1 - i2 * ((z - zc) ** 2 + x**2 + xc**2)))
+    return a, k2
+
+
+@nb.jit(nopython=True)
+def calc_e_k(
+    k2: float | np.ndarray,
+):
+    return ellipe_nb(k2), ellipk_nb(k2)
```
Comment (CoronelBuendia) [inline bluemira/magnetostatics/greens.py:304]: I believe the speed gains are potentially to be had in the combined use of AGMs to calculate E and K simultaneously...

Comment (geograham) [reply bluemira/magnetostatics/greens.py:304]: Hi, Sorry I missed this, that is true, I separated the issues into three - AGM speed gains are in this one: https://github.com/Fusion-Power-Plant-Framework/bluemira/issues/3273).

File: bluemira/magnetostatics/greens.py [hunk 2]
```diff
@@ -288,12 +278,44 @@ def greens_dpsi_dz(
...
+def calc_i1_i2(
+    a: float | np.ndarray,
+    k2: float | np.ndarray,
+    e: float | np.ndarray | None = None,
+    k: float | np.ndarray | None = None,
+):
+    if (e is None) or (k is None):
+        e, k = calc_e_k(k2)
+    i1 = k / a
+    i2 = e / (a**3 * (1 - k2))
```
Comment (je-cook) [inline bluemira/magnetostatics/greens.py:317]: Missed during the conversion
```suggestion
    i1 = 4 * k / a
    i2 = 4 * e / (a**3 * (1 - k2))
```

Comment (je-cook) [reply bluemira/magnetostatics/greens.py:317]: actually sometimes its there an sometimes its not

Comment (je-cook) [reply bluemira/magnetostatics/greens.py:317]: scratch this one

File: bluemira/magnetostatics/greens.py [hunk 3]
```diff
@@ -424,16 +446,12 @@ def greens_all(
...
-    a = np.hypot((x + xc), (z - zc))
-    k2 = 4 * x * xc / a**2
-    # Avoid NaN when coil on grid point
-    k2 = clip_nb(k2, GREENS_ZERO, 1.0 - GREENS_ZERO)
-    e, k = ellipe_nb(k2), ellipk_nb(k2)
-    i_1 = 4 * k / a
-    i_2 = 4 * e / (a**3 * (1 - k2))
+    a, k2 = calc_a_k2(xc, zc, x, z)
+    e, k = calc_e_k(k2)
+    i1, i2 = calc_i1_i2(a, k2, e, k)
```
Comment (je-cook) [inline bluemira/magnetostatics/greens.py:451]:
```suggestion
    i1, i2 = calc_i1_i2(a, k2, e, k)
    i1 *= 4
    i2 *= 4
```

**Review by je-cook — APPROVED** [review]

**Pull request merged**